google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/GapicSpannerRpc.java

```diff
@@ -102,18 +103,57 @@
 import com.google.spanner.v1.Session;
 import com.google.spanner.v1.Transaction;
 import io.grpc.Context;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executors;
 import java.util.concurrent.Future;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadFactory;
 import javax.annotation.Nullable;
 import org.threeten.bp.Duration;
 
 /** Implementation of Cloud Spanner remote calls using Gapic libraries. */
 public class GapicSpannerRpc implements SpannerRpc {
+  /**
+   * {@link ExecutorProvider} that keeps track of the executors that are created and shuts these
+   * down when the {@link SpannerRpc} is closed.
+   */
+  private static final class ManagedInstantiatingExecutorProvider implements ExecutorProvider {
+    private static final int DEFAULT_THREAD_COUNT = 4;
+    private final List<ScheduledExecutorService> executors = new LinkedList<>();
+    private final ThreadFactory threadFactory;
+
+    private ManagedInstantiatingExecutorProvider(ThreadFactory threadFactory) {
+      this.threadFactory = threadFactory;
+    }
+
+    @Override
+    public boolean shouldAutoClose() {
+      return false;
```

Comment:
Quick question - should this be configurable (the InstantiatingExecutorProvider returns true)? Or no, since we are now managing the closing ourselves? If this is true, it gets added to background resources in gax: https://github.com/googleapis/gax-java/blob/master/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java#L136. But not sure if this is necessary since we are now handling all shutdowns.

Reply:
It should not be configurable, but return false to ensure that we handle the shutdown ourselves. As you mentioned, setting this to true will cause the executors to be added to the background resources in gax, but will (currently?) not be shutdown by gax. To prevent unexpected regression if gax behavior should change in the future, I would say we should keep this returning false.

Reply:
Nevermind, sorry - this is only for the transport provider I realized :)